frame/support/src/traits/tasks.rs

```diff
@@ -0,0 +1,43 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Contains the [`Task`] trait, which defines a general-purpose way for defining and executing
+//! service work, and supporting types.
+
+use codec::FullCodec;
+use scale_info::TypeInfo;
+use sp_runtime::DispatchError;
+use sp_std::iter::Iterator;
+
+/// A general-purpose trait which defines a type of service work (i.e., work to performed by an
+/// off-chain worker) including methods for enumerating, validating, indexing, and running
+/// tasks of this type.
+pub trait Task: Sized + FullCodec + TypeInfo {
+	type Enumeration: Iterator<Item = Self>;
```

Comment:
so a new blocker is this line breaks object safety for the trait, meaning I can't write a method that accepts a `Trait` or `dyn Trait`. Any way we could do something like this without the usage of `Self` @bkchr?

Reply:
Actually if I add `'static` to `Enumeration` that works, but `EncodeLike` is also preventing it from being object safe :/

```
    |
341 | pub trait FullEncode: Encode + EncodeLike {}
    |                                ^^^^^^^^^^ the trait cannot be made into an object because it uses `Self` as a type parameter
```